crawl-ref/source/dgn-overview.cc

```diff
@@ -1094,3 +1064,164 @@ bool connected_branch_can_exist(branch_type br)
 
     return true;
 }
+
+static int _prompt_annotate_branch(level_id lid)
+{
+    // XXX I'd be grateful if someone could refactor this into something that's
+    // not black magic. The functionality doesn't have the be the exact same,
+    // so long as the list is of reasonable length.
+    // these 3 lines make a vector containing all shown branches.
+    // the current implementation shows all known branches.
+    vector<branch_type> brs;
+    for (branch_iterator it; it; ++it)
+        if (is_known_branch_id((**it).id))
+            brs.push_back(it->id);
+
+    while (true)
+    {
+        // Make the little overview
+        // (D) Dungeon        (T) Temple         (L) Lair           etc.
+        // at most 4 branches on 1 line
+        clear_messages();
+        int linec = 0;
+        string line;
+        for (branch_type br : brs)
+        {
+            if (linec == 4)
+            {
+                linec = 0;
+                mpr(line);
+                line = "";
+            }
+            line += make_stringf("(%c) %-14s ",
+                                 branches[br].travel_shortcut,
+                                 branches[br].shortname);
+            ++linec;
+        }
+        if (!line.empty())
+            mpr(line);
+
+        // tells you some special keys, but not all
+        mprf(MSGCH_PROMPT, "Annotate which branch? (. - %s, ? -help)",
+             lid.describe(false, true).c_str());
+
+        int keyin = get_ch();
+        switch (keyin)
+        {
+        CASE_ESCAPE
+            return ID_CANCEL;
+        case '?':
+            return ID_HELP;
+        case '\n': case '\r': case '.':
+            return ID_HERE;
+        case '<':
+            return ID_UP;
+        case '>':
+            return ID_DOWN;
+        case CONTROL('P'):
+            {
+                const branch_type parent = parent_branch(lid.branch);
+                if (parent < NUM_BRANCHES)
+                    return parent;
+            }
+            break;
+        default:
+            // Is this a branch hotkey?
+            for (branch_type br : brs)
+            {
+                if (toupper_safe(keyin) == branches[br].travel_shortcut)
+                    return br;
+            }
+            // Otherwise cancel
+            return ID_CANCEL;
+        }
+    }
+}
+
+void do_annotate()
+{
+    level_id lid  = level_id::current();
+    int branch = _prompt_annotate_branch(lid);
```

Comment:
these should be const if possible

Reply:
lid can't be const. I'll make branch constant since that's what you want